### ISSUE TITLE
Copy simultaniously: "file-unpacker: Fatal error: Cannot umount incoming directory (error type: Invalid argument)"

### ISSUE THREAD

**Issue by jamke** [QubesOS/qubes-issues#5950]
**Qubes OS version**
R4.0.
The dom0 is up to date from default repo on the date 2020.07.13. 

**Affected component(s) or functionality**
qvm-copy / qfile-unpacker

**Brief summary**
When copying files simultaneously - the message with an error pops up in target vm: "_qfile-unpacker: Fatal error: Cannot umount incoming directory (error type: Invalid argument)_" after last copy process finishes.

**To Reproduce**
1. Copy 5 files of ~300 MiB each from source-vm to target-vm via GUI commands.
2. Copy other 5 files of ~300 MiB each from source-vm to target-vm via GUI commands while the first copy is still ongoing.
Reproducible in 100% cases.

**Expected behavior**
Copy processes finished without any error messages.

**Actual behavior**
The first copy process finishes with no issues. The second process finishes and pops this window in target-vm with a color of target-vm:
> Caption: [target-vm] Error
> Text:    qfile-unpacker: Fatal error: Cannot umount incoming directory (error type: Invalid argument)
> Buttons: OK

**Additional context**
1. Probably it's the same with qvm-move as it's almost the same code plus remove of original files.
2. Probably both process some kind of "mount" the target directory QubesIncoming in target-vm and the first process to finish "umounts" it and for the second it's already "umount"-ed when it finished. Not sure what "mount/umount" means in this context, though.

**Solutions you've tried**
Copy files without simultaneous copy processes for the same target-vm.

**Related issues**
I searched for "qfile-unpacker" issues but did not find any with the same exact problem.

**Comment by marmarek** [comment]
>     2\. Probably both process some kind of "mount" the target directory QubesIncoming in target-vm and the first process to finish "umounts" it and for the second it's already "umount"-ed when it finished. Not sure what "mount/umount" means in this context, though.

Yes, this is exactly the case. qfile-unpacker do a bind-mount to avoid accidental directory move during copy.

**Comment by jamke** [comment]
Maybe bind-unmount part should be changed to check if it is still mounted or not or ignore unmount attempt fail.
But the proper solution is to make it somehow possible to mount twice (some kind of counter?) because otherwise the directory will be not bind-mounted during work of the second/third copy process.

**Comment by jamke** [comment]
Just checked. Affects 4.1.
```
qfile-unpacker: Fatal error: Cannot umount incoming directory (error type: Invalid argument)
```

**Comment by rustybird** [comment]
Looks like the problem is that `umount2()` propagates to all bind mount instances on that directory. This should be fixable by bind mounting with `MS_PRIVATE`, but doing so atomically also entails moving from `mount()` to `fsopen()` + `fsconfig()` + `fsmount()` from the new mount API in kernel 5.2+ (and glibc 2.36+, so it might be necessary to use `syscall()` directly)

WIP documentation for the new mount API: https://lore.kernel.org/linux-man/?q=fsopen

**Comment by rustybird** [comment]
Affects 4.2

**Comment by rustybird** [comment]
Affects 4.3

**Comment by datenbar** [comment]
Hi, I'd like to take this on if it's still up for grabs. 

From the thread, my understanding of the bug: `qfile-unpacker` bind-mounts the target-vm's incoming directory to guard against it being moved mid-transfer. When two copies run concurrently, both bind-mount the same directory but `umount2()` on that path unmounts *all* bind instances of it, not just the caller's, so the first copy to finish yanks the mount out from under the second one still in progress.

@rustybird's suggested fix direction (mount with `MS_PRIVATE` so each bind-mount is isolated, done atomically via `open_tree()` + `move_mount()`, since kernel 5.2+) makes sense to me. I'll start there unless someone's already partway through it.

Let me know if this is already claimed.

**Comment by andrewdavidwong** [comment]
@datenbar: This issue has not been claimed, and as far as I can tell, no one else is working on it. I'll assign this issue to you now.

**Comment by datenbar** [comment]
Update on this: I used Claude to build a standalone reproduction to test the proposed fix before pushing it. Two processes both bind-mounting the same directory onto itself, one finishes and unmounts while the other is still "mid-copy."

`open_tree(OPEN_TREE_CLONE)` + `mount_setattr(..., MS_PRIVATE)` + `move_mount()` did not fix it in my testing. The problem turned out to be that `umount2()` is purely path-based, it removes whatever mount is currently topmost at the target path. Since both processes' bind-mounts stack at the same path within the same shared mount namespace, whichever process calls `umount2()` first removes whatever's topmost at that moment, regardless of propagation type. `MS_PRIVATE` only stops an unmount event from propagating but doesn't change `umount2()` behavior.

What did fix it: `unshare(CLONE_NEWNS)`, called early in `main()` before the existing `mount()` call, no change to the mount call itself needed. Putting each process in its own mount namespace makes "topmost at this path" a private, per-process fact. Since `qfile-unpacker` is invoked fresh per copy, there's no cleanup concern the namespace is torn down automatically when the process exits.

Given that, is a plain `unshare(CLONE_NEWNS)` + the existing `mount()` call the right minimal fix, or is there a reason to still switch to `open_tree`/`mount_setattr`/`move_mount` on top of the `unshare()`?

The test harness I used is attached.

<details>

https://gist.github.com/datenbar/48a53cb2033b283db615d9faf2beba11

`unshare --user --map-root-user --mount --propagation unchanged ./race_test old`
`unshare --user --map-root-user --mount --propagation unchanged ./race_test new`

showed that in both the bind functions, adding unshare call gives similar results.

</details>

**Comment by marmarek** [comment]
But moving the mount to separate namespace will not prevent moving directories out of the incoming dir during copy anymore, no? So, that would defeat the original purpose of that mount.

**Comment by datenbar** [comment]
How about we drop `MNT_DETACH`, and call `umount2(path, 0)` in a loop until it errors, instead of once. A plain (non-lazy) unmount fails with `EBUSY` if the topmost mount is still busy and the child's existing `chroot(".")` already provides that busy reference. So a process that finishes and finds someone else's mount on top backs off cleanly (0 pops, EBUSY, exits). Whoever finishes last sweeps up any leftover non-busy layers below it in a loop, since nothing's protecting them anymore. Verified this holds with 2 concurrent processes and should in theory hold with n processes.

Test harness: [gist link](https://gist.github.com/datenbar/48a53cb2033b283db615d9faf2beba11)

Does this seem like the right direction, or is there a reason `MNT_DETACH` specifically was used originally that this might break?

**Comment by rustybird** [comment]
`MNT_DETACH` is to clean up the bind mount even if e.g. there's still a terminal open where the user had `cd`d into the directory during the copy process.

BTW this is also a limitation of the bind mount approach. If the user had already `cd`d into the directory before it was bind mounted, that shell session is not in the bind mount and hence not protected by it...

> The problem turned out to be that `umount2()` is purely path-based, it removes whatever mount is currently topmost at the target path.

Why is that a problem though? It should still result in the correct number of unmountings, even if an unpacker unmounts another unpacker's mount. (<- what a sentence)

Maybe it's racy inside the syscall? Two unpackers cleaning up at about the same time, both syscalls resolving the same topmost filesystem layer to detach, then the first one detaches it and the second one tries to and - is a it a no-op or an error? A (specific) error would be better, the unpacker could simply retry.

I was wondering anyway - could we just have one bind mount for however many parallel unpacker processes handling the same source VM, and `MNT_DETACH` it after the last one is done (non-racy, somehow)?

**Comment by datenbar** [comment]
Built a harness reproducing the real mount → fork/chroot → waitpid → umount2(MNT_DETACH) sequence with concurrent workers on the same path. Did reproduce fatal EINVAL failures in one environment, with the failing worker's own mount ID missing from /proc/self/mountinfo afterward, so it is kinda racy inside the syscall. However, I couldn't get this to reproduce consistently across different sandboxes (turned out I was unknowingly comparing runs across different filesystems/kernels), so I don't have a fully pinned-down mechanism I'd stand behind yet.

Given that, and since you'd mentioned the same idea, I think the right move is to have one shared bind mount per incoming_dir, refcounted, torn down only by whoever finishes last. I'm planning to use flock() on a lock file outside the bind-mounted subtree (so it's unaffected by the mount/unmount itself) to serialize the mount-if-first / unmount-if-last decision. Leaning toward each worker holding its own LOCK_SH for its lifetime, so a crashed/killed worker's reference is released automatically by the kernel.

Let me know if you see a problem with the flock() approach specifically, or if there's a reason it wouldn't fit here.

**Comment by datenbar** [comment]
One more thing before I start implementing: for the flock()-based refcount, I need a lock file path that's stable across concurrent invocations and outside the bind-mounted subtree.

Was thinking something like home_dir/QubesIncoming/.qfile-unpacker.lock.<remote_domain> one lock per source VM, created with O_CREAT if missing.

Since incoming_dir can also be passed directly via the legacy argv[2] path (not just the home_dir/remote_domain-derived one), wanted to check: is deriving the lock path from incoming_dir's parent directory safe/sane for both cases, or is there an existing convention elsewhere in the codebase I should follow instead? Also want to make sure I'm not missing a symlink/permissions concern given this sits in a directory the unprivileged user controls.

**Comment by rustybird** [comment]
Hmm how would this locking/refcounting scheme work? Sorry if it's an obvious design and I'm slow to get it. I've only been able to think of a different scheme where the primary unpacker is also responsible for unmounting.

> I need a lock file path that's [...] outside the bind-mounted subtree.

Does it really need to be? An open fd within the bind mount would delay fully unmounting it until the fd is closed, but that could still be okay with `MNT_DETACH`?

If you're not relying on opening the lock file with `O_EXCL` or on writing additional information into it, you could even flock an fd for the individual directory itself (opened with `O_RDONLY`, not `O_PATH`), rather than a file that would be left over after a crash. The directory is the same inode both before and after the bind mount, so it doesn't bother flock.

**Comment by datenbar** [comment]
The scheme works such that each worker holds a LOCK_SH on an fd for its entire lifetime (from before mounting until after unmounting). Multiple workers can hold LOCK_SH simultaneously. To check "am I first" (should I mount) or "am I last" (should I unmount), a worker attempts LOCK_EX | LOCK_NB on the same LOCK_SH fd, trying to upgrade it to exclusive to the same target: that only succeeds if no one currently holds any lock on it, so a successful exclusive acquisition is proof that I'm currently the only one here. If a worker crashes at any point, the kernel releases its LOCK_SH automatically when its fds close.

I like your idea of locking incoming_dir itself (opened O_RDONLY, not O_PATH) rather than a separate lock file since it's the same inode before and after the bind mount here, it avoids the whole path-derivation/symlink/legacy-argv question I raised, and there's nothing left over after a crash.

Whether holding that fd open through umount2(MNT_DETACH) is actually fine is something I would have to verify. I'll add a mode to my test that holds an O_RDONLY fd + LOCK_SH on the target directory across a real MNT_DETACH call and confirm nothing unexpected happens before I write this into the real code.

**Comment by rustybird** [comment]
Oh, I had assumed that we couldn't use lock upgrades/downgrades because the flock(2) manpage claims they aren't atomic. But that might be outdated? There's [a commit from 2015](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=267f1128583074b575b90a58de4dcb12dd25af96) saying Linus Torvalds complained about the non-atomicity, with the commit message then noting they "may want to revisit that at some point". And it looks like there's been a ton of changes to the code since then.

**Comment by datenbar** [comment]
Yes, the upgrade/downgrade still isn't atomic. But that does not affect the plan as much, since we're relying on the gurantee that only one caller can hold LOCK_EX.

Locking incoming_dir directly works well. Every worker tries LOCK_EX | LOCK_NB first. Whoever wins checks if it's already mounted and mounts if not, then downgrades to LOCK_SH for the rest of its lifetime. A worker that loses the race takes LOCK_SH and checks the mount state itself if it turns out nothing's actually mounted yet (the EX holder somehow died before finishing the mount), it drops its LOCK_SH and retries for LOCK_EX rather than assuming someone else has it handled (now this might have to loop? A couple of times? Or till we ensure mount is there?). 

On the way out, a worker releases LOCK_SH, requests for a blocking LOCK_EX, and only then checks and unmounts if it's genuinely the last one holding a reference.
Ran it under heavy concurrent load 64 workers, thousands of iterations, including worst-case timing with no delay between operations and it holds up cleanly, including with random SIGKILLs injected mid-run. This needs mount()/umount2() to use the absolute path, since under enough concurrent churn a worker's " . " can end up pointing at an already-torn-down mount.
